File: 39-combination-sum/Solution.cc
class Solution {
public:
    void bktk(const vector<int>& a, int target, int prev_i, int& sum, vector<int>& cur, vector<vector<int>>& r){
        if(sum == target){
            r.push_back(cur);
            return;
        }
        if(sum > target)
            return;
        int n = a.size();
        for(int i=prev_i; i<n; i++){
            sum += a[i];
            cur.push_back(a[i]);
            // Same number can be added multiple times, hence i
            bktk(a, target, i, sum, cur, r);
            // undo
            sum -= a[i];
            cur.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> r;
        vector<int> cur;
        // Sort the numbers
        sort(candidates.begin(), candidates.end());

        int sum = 0;
        bktk(candidates, target, 0, sum, cur, r);
        return r;
    }
};

File: 39-combination-sum/Solution.java
public class Solution {
    public void bktk(int[] candidates, int target, int prev_i, int sum, List<Integer> cur, List<List<Integer>> result ){
        if(sum == target){
            // add to the result set
            result.add(new ArrayList<Integer>(cur));
            return;
        }
        if(sum > target)
            return;
            
        for(int i=prev_i; i<candidates.length; i++){
            sum += candidates[i];
            cur.add(candidates[i]);

            // Each number can be used multiple times
            bktk(candidates, target, i, sum, cur, result);
            
            // undo
            sum -= candidates[i];
            cur.remove(cur.size()-1);
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList<Integer>();

        // Sort
        Arrays.sort(candidates);
        bktk(candidates, target, 0, 0, cur, result);
        return result;
    }
}